anvill/python/anvill/binja/bnprogram.py

```diff
@@ -78,6 +82,114 @@ def get_variable_impl(self, address):
 
         return BNVariable(bn_var, arch, address, var_type)
 
+    def _get_function_from_bnvariable(self, address, bn_var):
+        """Get the `Function` instance from the data variable of function
+        pointer type. Raise an `InvalidFunctionException` exception if
+        the data variable is not of function pointer type.
+        """
+
+        def _is_function_pointer(bn_var):
+            var_tinfo = bn_var.type
+            if (
+                var_tinfo.type_class == bn.TypeClass.PointerTypeClass
+                and var_tinfo.element_type.type_class == bn.TypeClass.FunctionTypeClass
+            ):
+                return True
+            return False
+
+        if bn_var is None:
+            return None
+
+        if not _is_function_pointer(bn_var):
+            return None
+
+        arch = self._arch
+        func_tinfo = bn_var.type.element_type
+
+        # TODO(akshayk): The type information may not have calling convention
+        # information. It does not recover the parameters and get the function
+        # in such case. A working solution could be handing the default calling
+        # convention for each architecture. This is in todo list.
+        if func_tinfo.calling_convention is None:
+            return
+
+        calling_conv = CallingConvention(arch, func_tinfo)
+
+        func_type = self.type_cache.get(func_tinfo)
+
+        index = 0
+        param_list = []
+        stack_offset = arch.stack_offset()
+        for var in func_tinfo.parameters:
+            var_type = var.type
+            reg_name = None
+
+            if var_type is None:
+                continue
+
+            arg_type = self.type_cache.get(var_type)
+
+            if (
+                bn.TypeClass.IntegerTypeClass == var_type.type_class
+                or bn.TypeClass.PointerTypeClass == var_type.type_class
+            ):
+                reg_name = calling_conv.next_int_arg_reg
+
+            elif bn.TypeClass.FloatTypeClass == var_type.type_class:
+                reg_name = calling_conv.next_float_arg_reg
+
+            elif bn.TypeClass.NamedTypeReferenceClass == var_type.type_class:
+                # The function paramater could be named alias of a float type.
+                if isinstance(arg_type, FloatingPointType):
+                    reg_name = calling_conv.next_float_arg_reg
+                else:
+                    reg_name = calling_conv.next_int_arg_reg
+
+            elif bn.TypeClass.VoidTypeClass == var_type.type_class:
+                reg_name = "invalid void"
+
+            else:
+                reg_name = None
+                raise AnvillException(
+                    "No variable type defined for function parameters"
+                )
+
+            if reg_name:
+                loc = Location()
+                loc.set_register(reg_name.upper())
```

Comment:
Instead of `reg_name.upper()`, you should use the `arch.register_name(reg_name)` method; that is why it is there. If yo usee other uses of `.upper()`, then please transition them to `arch.register_name(reg_name)`; using `.upper()` is not always the right approach.